[RA2 Ch3] Device Plugin operation and example (#2415)
cntt-n/CNTT: doc/ref_arch/kubernetes/chapters/chapter03.md

[Device Plugin Framework](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) advertises device hardware resources to kubelet with which vendors can implement plugins for devices that may require vendor-specific activation and life cycle management, and securely maps these devices to containers.

Figure 3-2 shows in four steps how device plugins operate on a Kubernetes node:

> * 1: During setup, the cluster administrator (more in [3.2.1.8 Operator Pattern](chapter03.md#3218-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different nodes, selects which devices to enable and deploys the associated device plugins.

Reviewer: ```suggestion
* 1: During setup, the cluster administrator (more in [3.2.5.1 Operator Pattern](chapter03.md#3251-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different nodes, selects which devices to enable and deploys the associated device plugins.
```
fixed

> * 1: During setup, the cluster administrator as human or operator in software (more in [3.2.1.8 Operator Pattern](chapter03.md#3218-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different machines, selects which devices to enable and deploys the associated device plugins.

Reviewer 2: @petorre @arkadykanevsky 
I don't think "human or software" is necessary
In other bullets, "node" is used
```suggestion
* 1: During setup, the cluster administrator (more in [3.2.1.8 Operator Pattern](chapter03.md#3218-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different nodes, selects which devices to enable and deploys the associated device plugins.
```

* 2: The plugin reports the devices it found on the node to the Kubelet device manager and starts its gRPC server to monitor the devices.
* 3: A user submits a pod specification (workload manifest file) requesting a certain type of device.

> * 4: Kubelet decides which node to schedule workload on and which device on the node to assign to the pod's containers.

Reviewer 3: nit-pick: the scheduler decides the node and kubelet handles the device allocation.
```suggestion
* 4: The scheduler determines a suitable node based on device availability and the local kubelet assigns a specific device to the pod's containers.
```

Author: > nit-pick: the scheduler decides the node and kubelet handles the device allocation.

@gkunz : Thank you!

<p align="center"><img src="../figures/Ch3_Figure_Device_Plugin_operation.png" alt="Device Plugin Operation" Title="Device Plugin Operation" width="50%"/></p>
<p align="center"><b>Figure 3-2:</b> Device Plugin Operation</p>

> An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin (such as [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)), to provision additional network interfaces for VFs (beyond the original network interface for Kubernetes control), and a SR-IOV CNI is required. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.

Reviewer: ```suggestion
An example of often used device plugin is the [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin), that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, the SR-IOV CNI is required, as well as a CNI multiplexer plugin (such as [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)), to provision additional secondary network interfaces for VFs (beyond the primary network interface). During pod creation, the plugin allocates a SR-IOV VF to a pod's network namespace using the VF information given by the meta plugin. On pod deletion, the CNI plugin releases the VF from the pod.
```
clarification

> An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, so is used to map VFs to scheduled pods. To use it needed are CNI meta plugin supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni), to have additional network interface for VFs beyond original network interface for Kubernetes control) and SR-IOV CNI (During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.)

Reviewer 2: typo
```suggestion
An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin, supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni), to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control) and SR-IOV CNI. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.
```

Figure 3-2 shows in four steps how device plugins operate on a Kubernetes node:
* 1: During setup, the cluster administrator (more in [3.2.5.1 Operator Pattern](chapter03.md#3251-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different nodes, selects which devices to enable and deploys the associated device plugins.
* 2: The plugin reports the devices it found on the node to the Kubelet device manager and starts its gRPC server to monitor the devices.
* 3: A user submits a pod specification (workload manifest file) requesting a certain type of device.
* 4: Kubelet decides which node to schedule workload on and which device on the node to assign to the pod's containers.

<p align="center"><img src="../figures/Ch3_Figure_Device_Plugin_operation.png" alt="Device Plugin Operation" Title="Device Plugin Operation" width="50%"/></p>
<p align="center"><b>Figure 3-2:</b> Device Plugin Operation</p>

> An example of often used device plugin is the [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin), that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, the SR-IOV CNI is required, as well as a CNI multiplexer plugin (such as [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)), to provision additional secondary network interfaces for VFs (beyond the primary network interface). During pod creation, the plugin allocates a SR-IOV VF to a pod's network namespace using the VF information given by the meta plugin. On pod deletion, the CNI plugin releases the VF from the pod.

Reviewer 2: In the last sentence can we clarify which plugin as the paragraph references the SR-IOV and the multiplexer plugins.

Author: > In the last sentence can we clarify which plugin as the paragraph references the SR-IOV and the multiplexer plugins.

@pgoyal01 : Please see if current text clarifies it.

Reviewer 2: @petorre Thanks.

> An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin, supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)) and SR-IOV CNI, to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control), are required. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.

Reviewer 2: ```suggestion
An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin (such as [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)), to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control), and a SR-IOV CNI is required. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.
```

<p align="center"><img src="../figures/Ch3_Figure_Device_Plugin_operation.png" alt="Device Plugin Operation" Title="Device Plugin Operation" width="50%"/></p>
<p align="center"><b>Figure 3-2:</b> Device Plugin Operation</p>

> An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin, supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni), to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control) and SR-IOV CNI. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.

Reviewer 4: ```suggestion
An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin, supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)) and SR-IOV CNI, to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control), are required. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.
```
